File: src/engine/command.ts
import { CAPTURE_BYTE_LIMIT, CaptureLimitError } from "../caps";
import { validateHerdrInvocation } from "../host";
import { renderScalar, substituteParams, substituteText } from "../workflow/grammar";
import type { PaneOpen, ShellName, StepAction, TemplateNamespace } from "../workflow/grammar";
import type { CommandResult, StepFailureDetails } from "../workflow/results";
import {
  dispatchFailure,
  errorText,
  readTruncated,
  type StepFrame,
  type StepOutcome,
} from "./contract";
import { placeCommandPane, resolvePaneOpen } from "./pane";

export async function herdrStep(frame: StepFrame): Promise<StepOutcome> {
  const action = frame.step.action;
  if (action.kind !== "herdr") return { ok: false, error: "internal: not a herdr step" };
  const params = substituteParams(action.params, frame.values) ?? {};
  const invalid = validateHerdrInvocation(action.method, params);
  if (invalid) return { ok: false, error: invalid, details: { method: action.method } };
  try {
    const result = await frame.opts.deps.herdrCall(action.method, params);
    return { ok: true, result, ...(readTruncated(result) ? { truncated: true } : {}) };
  } catch (error) {
    const failure = dispatchFailure(`herdr ${action.method}`, error);
    return failure.ok ? failure : { ...failure, details: { method: action.method } };
  }
}

type CommandAction = Extract<StepAction, { kind: "run" }>;

type CaptureBudget = {
  source: string;
  limit: number;
  total: number;
  onOverflow: () => void;
};

async function readStreamAgainstBudget(
  stream: ReadableStream<Uint8Array> | null,
  budget?: CaptureBudget,
): Promise<string> {
  if (!stream) return "";
  if (!budget) return new Response(stream).text();
  const reader = stream.getReader();
  const chunks: Uint8Array[] = [];
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    if (!value?.byteLength) continue;
    budget.total += value.byteLength;
    if (budget.total > budget.limit) {
      budget.onOverflow();
      try {
        await reader.cancel();
      } catch {
        /* already closed */
      }
      throw new CaptureLimitError(budget.source, budget.total, budget.limit);
    }
    chunks.push(value);
  }
  return Buffer.concat(chunks).toString("utf8");
}

export function shellArgv(command: string, shell: ShellName = "sh"): string[] {
  switch (shell) {
    case "sh":
      return ["sh", "-c", command];
    case "bash":
      return ["bash", "-c", command];
    case "zsh":
      return ["zsh", "-c", command];
    case "pwsh":
      return ["pwsh", "-NoProfile", "-Command", command];
    case "powershell":
      return ["powershell", "-NoProfile", "-Command", command];
    case "cmd":
      return ["cmd", "/c", command];
  }
}

export function killSpawn(proc: { pid: number; kill: () => void }): void {
  try {
    process.kill(-proc.pid, "SIGKILL");
  } catch {
    try {
      proc.kill();
    } catch {
      /* already dead */
    }
  }
}

export type CaptureResult = {
  timedOut: boolean;
  exitCode: number;
  stdout: string;
  stderr: string;
  timeoutMs: number;
};

/** `timeoutMs` omitted means no workflow timeout; process completion still blocks. */
export async function spawnCapture(
  argv: string[],
  opts: {
    cwd: string;
    stdin?: string;
    env?: NodeJS.ProcessEnv;
    timeoutMs?: number;
    maxCaptureBytes?: { source: string };
  },
): Promise<CaptureResult> {
  const timeoutMs = opts.timeoutMs ?? 0;
  const proc = Bun.spawn(argv, {
    cwd: opts.cwd,
    stdin: "pipe",
    stdout: "pipe",
    stderr: "pipe",
    env: opts.env,
    detached: true,
  });
  if (opts.stdin !== undefined) proc.stdin.write(opts.stdin);
  proc.stdin.end();

  let timedOut = false;
  const timer =
    timeoutMs > 0
      ? setTimeout(() => {
          timedOut = true;
          killSpawn(proc);
        }, timeoutMs)
      : undefined;

  const budget = opts.maxCaptureBytes
    ? {
        source: opts.maxCaptureBytes.source,
        limit: CAPTURE_BYTE_LIMIT,
        total: 0,
        onOverflow: () => killSpawn(proc),
      }
    : undefined;

  try {
    const [stdout, stderr, exitCode] = await Promise.all([
      readStreamAgainstBudget(proc.stdout, budget),
      readStreamAgainstBudget(proc.stderr, budget),
      proc.exited,
    ]);
    clearTimeout(timer);
    return { timedOut, exitCode: exitCode ?? 1, stdout, stderr, timeoutMs };
  } catch (error) {
    clearTimeout(timer);
    killSpawn(proc);
    throw error;
  }
}

type CommandOutcome = {
  ok: boolean;
  stdout: string;
  stderr: string;
  exitCode: number;
  timedOut: boolean;
  failed: boolean;
};

function captureResult(r: CaptureResult, successCodes: number[] = [0]): CommandOutcome {
  const accepted = !r.timedOut && successCodes.includes(r.exitCode);
  const failed = !accepted;
  const stderr = r.timedOut && !r.stderr ? `timed out after ${r.timeoutMs / 1000}s` : r.stderr;
  return {
    ok: accepted,
    stdout: r.stdout,
    stderr,
    exitCode: r.exitCode,
    timedOut: r.timedOut,
    failed,
  };
}

const COMMAND_CAPTURE_SOURCE = "command output";

export async function runShellStep(
  command: string,
  opts: {
    cwd: string;
    stdin?: string;
    env?: NodeJS.ProcessEnv;
    timeoutMs?: number;
    shell?: ShellName;
    successCodes?: number[];
  },
): Promise<CommandOutcome> {
  return captureResult(
    await spawnCapture(shellArgv(command, opts.shell), {
      cwd: opts.cwd,
      stdin: opts.stdin,
      env: opts.env,
      timeoutMs: opts.timeoutMs,
      maxCaptureBytes: { source: COMMAND_CAPTURE_SOURCE },
    }),
    opts.successCodes,
  );
}

export async function runArgvStep(
  argv: string[],
  opts: {
    cwd: string;
    env?: NodeJS.ProcessEnv;
    timeoutMs?: number;
    successCodes?: number[];
  },
): Promise<CommandOutcome> {
  return captureResult(
    await spawnCapture(argv, {
      cwd: opts.cwd,
      env: opts.env,
      timeoutMs: opts.timeoutMs,
      maxCaptureBytes: { source: COMMAND_CAPTURE_SOURCE },
    }),
    opts.successCodes,
  );
}

export function buildHwfEnv(inputs: Record<string, unknown>): Record<string, string> {
  return Object.fromEntries(
    Object.entries(inputs).map(([name, value]) => [`HWF_${name}`, renderScalar(value)]),
  );
}

const RESERVED_ENV_RE = /^hwf_/i;

type StepEnv = { ok: true; env: Record<string, string> } | { ok: false; error: string };

function stepEnvValues(env: Record<string, string> | undefined, ns: TemplateNamespace): StepEnv {
  const out: Record<string, string> = {};
  for (const [key, value] of Object.entries(env ?? {})) {
    if (RESERVED_ENV_RE.test(key)) {
      return { ok: false, error: `env key '${key}' uses the reserved HWF_ prefix` };
    }
    out[key] = substituteText(value, ns);
  }
  return { ok: true, env: out };
}

/** Runner-generated HWF values replace inherited collisions; explicit `env:` wins over both. */
export function mergeStepEnv(
  inherited: NodeJS.ProcessEnv,
  hwf: Record<string, string>,
  stepEnv: Record<string, string>,
): NodeJS.ProcessEnv {
  return { ...inherited, ...hwf, ...stepEnv };
}

function commandArgv(action: CommandAction, ns: TemplateNamespace): string[] {
  const payload = action.payload;
  if (payload.form === "argv") return payload.argv.map((el) => substituteText(el, ns));
  return shellArgv(payload.command, payload.shell);
}

function bindCommandResult(frame: StepFrame, outcome: CommandOutcome): void {
  if (!frame.step.id) return;
  const result: CommandResult = {
    stdout: outcome.stdout,
    stderr: outcome.stderr,
    exit_code: outcome.exitCode,
    failed: outcome.failed,
  };
  frame.values.steps[frame.step.id] = result;
}

function commandFailure(outcome: CommandOutcome): Extract<StepOutcome, { ok: false }> {
  const detail = outcome.stderr.trim() || outcome.stdout.trim().slice(-500);
  const details: StepFailureDetails = {
    stdout: outcome.stdout,
    stderr: outcome.stderr,
    exit_code: outcome.exitCode,
  };
  return {
    ok: false,
    error: detail || `exit ${outcome.exitCode}`,
    details,
  };
}

async function localCommand(
  frame: StepFrame,
  action: CommandAction,
  cwd: string,
  env: NodeJS.ProcessEnv,
): Promise<StepOutcome> {
  const payload = action.payload;
  const successCodes = action.successCodes ?? [0];
  let outcome: CommandOutcome;
  try {
    outcome =
      payload.form === "argv"
        ? await runArgvStep(commandArgv(action, frame.values), {
            cwd,
            env,
            timeoutMs: action.timeoutMs,
            successCodes,
          })
        : await runShellStep(payload.command, {
            cwd,
            env,
            timeoutMs: action.timeoutMs,
            shell: payload.shell,
            successCodes,
          });
  } catch (error) {
    if (error instanceof CaptureLimitError) {
      return { ok: false, error: error.message, hardFailure: true };
    }
    return { ok: false, error: `run: ${errorText(error)}`, hardFailure: true };
  }
  if (outcome.stderr) frame.opts.onStderr?.(outcome.stderr);
  if (outcome.timedOut) return { ...commandFailure(outcome), hardFailure: true };
  bindCommandResult(frame, outcome);
  return outcome.failed ? commandFailure(outcome) : { ok: true };
}

const READY_LINES = 80;

async function placedCommand(
  frame: StepFrame,
  action: CommandAction,
  cwd: string,
  paneEnv: Record<string, string>,
): Promise<StepOutcome> {
  const pane = action.pane;
  if (!pane) return { ok: false, error: "run: background and ready_when require pane:" };
  const sub = (text?: string) =>
    text === undefined ? undefined : substituteText(text, frame.values);
  let open: PaneOpen;
  try {
    open = resolvePaneOpen(pane.open, frame.values);
  } catch (error) {
    return { ok: false, error: errorText(error) };
  }
  const placed = await placeCommandPane({
    open,
    anchor: sub(pane.anchor),
    workspace: sub(pane.workspace),
    size: pane.size,
    focus: pane.focus ?? action.background !== true,
    cwd,
    env: paneEnv,
    label: sub(pane.name) ?? frame.step.id ?? "hwf-run",
    argv: commandArgv(action, frame.values),
    deps: frame.opts.deps,
    invocation: frame.opts.ctx,
  });
  if (action.background === true) return { ok: true, launched: true };
  if (action.readyWhen === undefined || action.timeoutMs === undefined) {
    return { ok: false, error: "run: placed foreground run requires ready_when and timeout" };
  }
  const waited = await frame.opts.deps.herdrCall("pane.wait_for_output", {
    pane_id: placed.pane_id,
    source: "recent",
    lines: READY_LINES,
    strip_ansi: true,
    match: { type: "regex", value: action.readyWhen },
    timeout_ms: action.timeoutMs,
  });
  return {
    ok: true,
    result: { ...waited, ...placed },
    ...(readTruncated(waited) ? { truncated: true } : {}),
  };
}

export async function shellStep(
  frame: StepFrame & { env: NodeJS.ProcessEnv },
): Promise<StepOutcome> {
  const action = frame.step.action;
  if (action.kind !== "run") return { ok: false, error: "internal: not a run step" };
  const stepEnv = stepEnvValues(action.env, frame.values);
  if (!stepEnv.ok) return { ok: false, error: stepEnv.error };
  const hwf = buildHwfEnv(frame.values.inputs);
  const cwd =
    action.cwd !== undefined ? substituteText(action.cwd, frame.values) : frame.opts.ctx.cwd;
  if (action.pane || action.background === true || action.readyWhen !== undefined) {
    try {
      return await placedCommand(frame, action, cwd, { ...hwf, ...stepEnv.env });
    } catch (error) {
      return dispatchFailure("run", error);
    }
  }
  return localCommand(frame, action, cwd, mergeStepEnv(frame.env, hwf, stepEnv.env));
}

File: internal/engine/command.go
package engine

import (
	"fmt"
	"maps"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"

	"github.com/aorumbayev/herdr-workflows/internal/caps"
	"github.com/aorumbayev/herdr-workflows/internal/host"
	"github.com/aorumbayev/herdr-workflows/internal/workflow"
)

func ShellArgv(command, shell string) []string {
	switch shell {
	case "sh":
		return []string{"sh", "-c", command}
	case "bash":
		return []string{"bash", "-c", command}
	case "zsh":
		return []string{"zsh", "-c", command}
	case "pwsh":
		return []string{"pwsh", "-NoProfile", "-Command", command}
	case "powershell":
		return []string{"powershell", "-NoProfile", "-Command", command}
	case "cmd":
		return []string{"cmd", "/c", command}
	default:
		return []string{"sh", "-c", command}
	}
}

func NativeProcessTree(goos string) bool {
	return goos == "linux" || goos == "darwin"
}

type CommandOutcome struct {
	OK       bool
	Failed   bool
	TimedOut bool
	Stdout   string
	Stderr   string
	ExitCode int
}

type ShellStepOpts struct {
	Cwd          string
	Stdin        *string
	Env          []string
	TimeoutMs    int
	Shell        string
	SuccessCodes []int
}

type ArgvStepOpts struct {
	Cwd          string
	Env          []string
	TimeoutMs    int
	SuccessCodes []int
}

func formatTimeoutSeconds(ms int) string {
	seconds := float64(ms) / 1000.0
	return strconv.FormatFloat(seconds, 'f', -1, 64)
}

func captureResult(r caps.SpawnResult, successCodes []int) CommandOutcome {
	if len(successCodes) == 0 {
		successCodes = []int{0}
	}

	accepted := !r.TimedOut && slices.Contains(successCodes, r.ExitCode)
	stderr := r.Stderr
	if r.TimedOut && stderr == "" {
		stderr = fmt.Sprintf("timed out after %ss", formatTimeoutSeconds(r.TimeoutMs))
	}

	return CommandOutcome{
		OK:       accepted,
		Failed:   !accepted,
		TimedOut: r.TimedOut,
		Stdout:   r.Stdout,
		Stderr:   stderr,
		ExitCode: r.ExitCode,
	}
}

func RunShellStep(command string, opts ShellStepOpts) (CommandOutcome, error) {
	argv := ShellArgv(command, opts.Shell)

	captureOpts := caps.SpawnOpts{
		Cwd:              opts.Cwd,
		Stdin:            opts.Stdin,
		Env:              opts.Env,
		TimeoutMs:        opts.TimeoutMs,
		MaxCaptureSource: "command output",
	}

	result, err := caps.Spawn(argv, captureOpts)
	if err != nil {
		return CommandOutcome{}, err
	}

	return captureResult(result, opts.SuccessCodes), nil
}

func RunArgvStep(argv []string, opts ArgvStepOpts) (CommandOutcome, error) {
	captureOpts := caps.SpawnOpts{
		Cwd:              opts.Cwd,
		Env:              opts.Env,
		TimeoutMs:        opts.TimeoutMs,
		MaxCaptureSource: "command output",
	}

	result, err := caps.Spawn(argv, captureOpts)
	if err != nil {
		return CommandOutcome{}, err
	}

	return captureResult(result, opts.SuccessCodes), nil
}

func BuildHwfEnv(inputs map[string]any) map[string]string {
	result := make(map[string]string)
	for name, value := range inputs {
		result[fmt.Sprintf("HWF_%s", name)] = workflow.RenderScalar(value)
	}
	return result
}

func MergeStepEnv(inherited []string, hwf, stepEnv map[string]string) []string {
	envMap := make(map[string]string)

	for _, kv := range inherited {
		parts := strings.SplitN(kv, "=", 2)
		if len(parts) == 2 {
			envMap[parts[0]] = parts[1]
		}
	}

	for key, value := range hwf {
		envMap[key] = value
	}

	for key, value := range stepEnv {
		envMap[key] = value
	}

	result := make([]string, 0, len(envMap))
	keys := slices.Sorted(maps.Keys(envMap))
	for _, key := range keys {
		result = append(result, fmt.Sprintf("%s=%s", key, envMap[key]))
	}

	return result
}

func asAction[T any](action workflow.Action) (*T, bool) {
	if a, ok := any(action).(*T); ok {
		return a, true
	}
	if a, ok := any(action).(T); ok {
		return &a, true
	}
	return nil, false
}

func HerdrStep(frame StepFrame) (StepOutcome, error) {
	action, ok := asAction[workflow.HerdrAction](frame.Step.Action)
	if !ok {
		return StepOutcome{OK: false, Error: "internal: not a herdr step"}, nil
	}

	params := workflow.SubstituteParams(action.Params, frame.Values)
	if params == nil {
		params = map[string]any{}
	}

	invalid := host.ValidateHerdrInvocation(action.Method, params, workflow.IsWholeValueTemplate)
	if invalid != nil {
		return StepOutcome{
			OK:      false,
			Error:   invalid.Error(),
			Details: map[string]any{"method": action.Method},
		}, nil
	}

	result, err := frame.Opts.Deps.HerdrCall(action.Method, params)
	if err != nil {
		return DispatchFailure("herdr "+action.Method, err), nil
	}

	outcome := StepOutcome{
		OK:     true,
		Result: result,
	}
	if ReadTruncated(result) {
		outcome.Truncated = true
	}
	return outcome, nil
}

func bindCommandResult(frame StepFrame, outcome CommandOutcome) {
	if frame.Step.ID == "" {
		return
	}

	frame.Values.Steps[frame.Step.ID] = map[string]any{
		"stdout":    outcome.Stdout,
		"stderr":    outcome.Stderr,
		"exit_code": outcome.ExitCode,
		"failed":    outcome.Failed,
	}
}

func commandFailure(outcome CommandOutcome) StepOutcome {
	detail := strings.TrimSpace(outcome.Stderr)
	if detail == "" {
		trimmedStdout := strings.TrimSpace(outcome.Stdout)
		if runes := []rune(trimmedStdout); len(runes) > 500 {
			detail = string(runes[len(runes)-500:])
		} else {
			detail = trimmedStdout
		}
	}
	if detail == "" {
		detail = fmt.Sprintf("exit %d", outcome.ExitCode)
	}

	details := map[string]any{
		"stdout":    outcome.Stdout,
		"stderr":    outcome.Stderr,
		"exit_code": outcome.ExitCode,
	}

	return StepOutcome{
		OK:      false,
		Error:   detail,
		Details: details,
	}
}

func commandArgv(action *workflow.RunAction, ns workflow.TemplateNamespace) []string {
	if action.Payload.IsArgv() {
		argv := make([]string, len(action.Payload.Argv))
		for i, el := range action.Payload.Argv {
			argv[i] = workflow.SubstituteText(el, ns)
		}
		return argv
	}
	return ShellArgv(action.Payload.Command, action.Payload.Shell)
}

func stepEnvValues(env map[string]string, ns workflow.TemplateNamespace) (map[string]string, error) {
	out := make(map[string]string)
	reserved := regexp.MustCompile(`^[Hh][Ww][Ff]_`)

	for key, value := range env {
		if reserved.MatchString(key) {
			return nil, fmt.Errorf("env key '%s' uses the reserved HWF_ prefix", key)
		}
		out[key] = workflow.SubstituteText(value, ns)
	}
	return out, nil
}

func localCommand(
	frame StepFrame,
	action *workflow.RunAction,
	cwd string,
	env []string,
) StepOutcome {
	payload := action.Payload
	successCodes := action.SuccessCodes
	if len(successCodes) == 0 {
		successCodes = []int{0}
	}

	var outcome CommandOutcome
	var err error

	if payload.IsArgv() {
		outcome, err = RunArgvStep(commandArgv(action, frame.Values), ArgvStepOpts{
			Cwd:          cwd,
			Env:          env,
			TimeoutMs:    int(action.Timeout.Milliseconds()),
			SuccessCodes: successCodes,
		})
	} else {
		outcome, err = RunShellStep(payload.Command, ShellStepOpts{
			Cwd:          cwd,
			Env:          env,
			TimeoutMs:    int(action.Timeout.Milliseconds()),
			Shell:        payload.Shell,
			SuccessCodes: successCodes,
		})
	}

	if err != nil {
		if _, ok := err.(*caps.CaptureLimitError); ok {
			return StepOutcome{OK: false, Error: err.Error(), HardFailure: true}
		}
		return StepOutcome{OK: false, Error: "run: " + ErrorText(err), HardFailure: true}
	}

	if outcome.Stderr != "" {
		if frame.Opts.OnStderr != nil {
			frame.Opts.OnStderr(outcome.Stderr)
		}
	}

	if outcome.TimedOut {
		fail := commandFailure(outcome)
		fail.HardFailure = true
		return fail
	}

	bindCommandResult(frame, outcome)

	if outcome.Failed {
		return commandFailure(outcome)
	}

	return StepOutcome{OK: true}
}

func placedCommand(
	frame StepFrame,
	action *workflow.RunAction,
	cwd string,
	paneEnv map[string]string,
) StepOutcome {
	pane := action.Pane
	if pane == nil {
		return StepOutcome{OK: false, Error: "run: background and ready_when require pane:"}
	}

	sub := func(text string) string {
		return workflow.SubstituteText(text, frame.Values)
	}

	open, err := ResolvePaneOpen(pane.Open, frame.Values)
	if err != nil {
		return StepOutcome{OK: false, Error: ErrorText(err)}
	}

	anchor := ""
	if pane.Anchor != "" {
		anchor = sub(pane.Anchor)
	}

	workspace := ""
	if pane.Workspace != "" {
		workspace = sub(pane.Workspace)
	}

	label := ResolvePaneLabel(pane.Name, frame.Values, frame.Step.ID)

	focus := !action.Background
	if pane.Focus != nil {
		focus = *pane.Focus
	}

	placed, err := PlaceCommandPane(PlaceOpts{
		Open:       open,
		Anchor:     anchor,
		Workspace:  workspace,
		Size:       pane.Size,
		Focus:      focus,
		Cwd:        cwd,
		Env:        paneEnv,
		Label:      label,
		Argv:       commandArgv(action, frame.Values),
		Deps:       frame.Opts.Deps,
		Invocation: frame.Opts.Ctx,
	})
	if err != nil {
		return DispatchFailure("run", err)
	}

	if action.Background {
		return StepOutcome{OK: true, Launched: true}
	}

	if action.ReadyWhen == "" || action.Timeout == 0 {
		return StepOutcome{OK: false, Error: "run: placed foreground run requires ready_when and timeout"}
	}

	waitParams := map[string]any{
		"pane_id":    placed.PaneID,
		"source":     "recent",
		"lines":      80,
		"strip_ansi": true,
		"match": map[string]any{
			"type":  "regex",
			"value": action.ReadyWhen,
		},
		"timeout_ms": int(action.Timeout.Milliseconds()),
	}

	waited, err := frame.Opts.Deps.HerdrCall("pane.wait_for_output", waitParams)
	if err != nil {
		return DispatchFailure("run", err)
	}

	result := maps.Clone(waited)
	if result == nil {
		result = map[string]any{}
	}
	result["pane_id"] = placed.PaneID
	result["tab_id"] = placed.TabID
	result["workspace_id"] = placed.WorkspaceID

	outcome := StepOutcome{OK: true, Result: result}
	if ReadTruncated(result) {
		outcome.Truncated = true
	}
	return outcome
}

func ShellStep(frame StepFrame) (StepOutcome, error) {
	action, ok := asAction[workflow.RunAction](frame.Step.Action)
	if !ok {
		return StepOutcome{OK: false, Error: "internal: not a run step"}, nil
	}

	stepEnv, err := stepEnvValues(action.Env, frame.Values)
	if err != nil {
		return StepOutcome{OK: false, Error: err.Error()}, nil
	}

	hwf := mergeEnv(BuildHwfEnv(frame.Values.Inputs), runContextEnv(frame.Opts))

	cwd := frame.Opts.Ctx.Cwd
	if action.Cwd != "" {
		cwd = workflow.SubstituteText(action.Cwd, frame.Values)
	}

	if action.Pane != nil || action.Background || action.ReadyWhen != "" {
		return placedCommand(frame, action, cwd, mergeEnv(hwf, stepEnv)), nil
	}

	env := frame.Opts.Env
	if env == nil {
		env = os.Environ()
	}

	mergedEnv := MergeStepEnv(env, hwf, stepEnv)
	return localCommand(frame, action, cwd, mergedEnv), nil
}

func runContextEnv(opts StepRunOpts) map[string]string {
	out := map[string]string{}
	if opts.RunID != "" {
		out["HWF_RUN_ID"] = opts.RunID
	}
	if opts.Name != "" {
		out["HWF_WORKFLOW"] = opts.Name
	}
	if opts.RepoRoot != "" {
		out["HWF_CHECKOUT_ROOT"] = opts.RepoRoot
	}
	return out
}
